src/uct/cuda/cuda_copy/cuda_copy_ep.c

```diff
@@ -148,15 +183,29 @@ uct_cuda_copy_ctx_rsc_get(uct_cuda_copy_iface_t *iface, CUdevice *cuda_device_p,
     ucs_status_t status;
     uct_cuda_ctx_rsc_t *ctx_rsc;
 
-    result = uct_cuda_base_ctx_get_id(NULL, &ctx_id);
-    if (ucs_likely(result == CUDA_SUCCESS)) {
-        /* If there is a current context, the CU_DEVICE_INVALID is returned in
-           cuda_device_p */
-        cuda_device = CU_DEVICE_INVALID;
+    /* If valid sys_dev is provided we need to retain and push primary context
+     * of the device. This is because of some limitation when using VMM and
+     * cuMemcpyAsync - the current context should match the device VMM has
+     * access to. */
+    status = uct_cuda_base_get_cuda_device(sys_dev, &cuda_device);
+    if (status == UCS_OK) {
+        status = uct_cuda_copy_push_ctx(cuda_device, 0, UCS_LOG_LEVEL_ERROR);
```

Comment:
I think the error handling is not valid now.
We can push the context, without retaining the primary context. Then we may release the primary context after failure of `uct_cuda_base_ctx_rsc_get`.

Reply:
can you resolve pls?